src/op_mode/ntp.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2024 VyOS maintainers and contributors
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 or later as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import sys
+import typing
+
+import vyos.opmode
+from vyos.configquery import ConfigTreeQuery
+from vyos.utils.process import cmd
+
+def _get_formatted_output(output: str) -> dict:
+
+    # Initialize dictionary keys to match output of
+    # chrony commmand being passed into this function.
+    # From some commands, the -c switch outputs more
+    # parameters, make sure to include them all below.
+    # See to chronyc(1) for definition of key variables
+    match output:
+        case "chronyc -c 'activity'":
+            keys: list = [
+            'sources_online',
+            'sources_offline',
+            'sources_doing_burst_return_online',
+            'sources_doing_burst_return_offline',
+            'sources_with_unknown_address'
+            ]
+
+        case "chronyc -c 'sources'":
+            keys: list = [
+            'm',
+            's',
+            'name_ip_address',
+            'stratum',
+            'poll',
+            'reach',
+            'last_rx',
+            'last_sample_adj_offset',
+            'last_sample_mes_offset',
+            'last_sample_est_error'
+            ]
+
+        case "chronyc -c 'sourcestats'":
+            keys: list = [
+            'name_ip_address',
+            'np',
+            'nr',
+            'span',
+            'frequency',
+            'freq_skew',
+            'offset',
+            'std_dev'
+            ]
+
+        case "chronyc -c 'tracking'":
+            keys: list = [
+            'ref_id',
+            'ref_id_name',
+            'stratum',
+            'ref_time',
+            'system_time',
+            'last_offset',
+            'rms_offset',
+            'frequency',
+            'residual_freq',
+            'skew',
+            'root_delay',
+            'root_dispersion',
+            'update_interval',
+            'leap_status'
+            ]
+
+        case _:
+            raise ValueError(f"Raw mode: of {output} is not implemented")
+
+    # Get command line output and apply regex to remove
+    # commas and newlines characters
+    values = cmd(output)
+    values = list(re.split(r'[,\n]', values))
```

Comment:
Chrony's output is CSV and there's a CSV library for Python (`import csv`). Using the library will make the code simpler and more tolerant to possible format changes.

Parsing CSVs with regexes has multiple pitfalls and the `csv` module already accounts for stuff like quoted strings, etc.

Reply:
Noted will change to use csv.reader()

Reply:
@dmbaturin your feedback has been incorporated.